daemon/protocol/python/scoot/scoot.py

```diff
@@ -0,0 +1,199 @@
+                          Default: 0
+  --all                   Return the status of all the runs not just finished runs.
+
+"""
+import docopt
+import os
+import sys
+import re
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '../../protocol'))
+import client_lib as proto
+
+
+def display_statuses(statuses):
+  for status in statuses:
+    print("\nRunId:{0},\n\tState:{1},\n\tExitCode:{2},\n\tError:{3},\n\tSnapshot:{4}\n".format(status.run_id, proto.display_state(status.state), status.exit_code, status.error, status.snapshot_id))
+
+
+def snapshot_create(cmd):
+    #verify args
+  if not cmd['<srcDir>']:
+    raise docopt.DocoptExit("Must provide srcDir!")
+  
+  #run the command
+  try:
+    sid = proto.create_snapshot(cmd['<srcDir>'])
+    # print the output
+    print("snapshot id = {0}".format(sid))
+  #handle errors running the command
+  except proto.ScootException as e:
+    if re.search("Not started", e) > 0 or re.search("UNAVAILABLE", e) > 0:
+      raise docopt.DocoptExit("Scoot Daemon is not running!\n")
+    raise docopt.DocoptExit("create snapshot error: '{0}'.".format(str(e)))       
+
+
+def snapshot_checkout(cmd):
+    #verify args
+  if not cmd['<snapshotId>']:
+    raise docopt.DocoptExit("Must supply a snapshot id.")
+  if not cmd["<destDir>"]:
+    raise docopt.DocoptExit("Must supply a destination directory.")
+  #run the command
+  try:
+    r = proto.checkout_snapshot(snapshot_id=cmd['<snapshotId>'], dirpath=cmd["<destDir>"])
+  #handle errors running the command
+  except Exception as e:
+    raise docopt.DocoptExit("snapshot checkout error: '{0}'.".format(str(e)))
+  
+  
+def run(cmd):
+  #verfiy args
+  timeout = cmd['--timeout']
+  if not timeout:
+    timeout_ns = int(1e9)
+  else:
+    try:
+      timeout_ns = int(int(timeout) * 1e9)
+    except Exception as e:
+      raise docopt.DocoptExit("invalid value for timeout, must be an integer or decimal number.")
+  snapshotId = cmd['--snapshotId']
+  args = cmd['<command>']
+
+  #run the command
+  try:
+    runId = proto.run(snapshot_id=snapshotId, argv=args, timeout_ns=timeout_ns)
+    #print the output
+    print("run id = {0}".format(runId))
+  #handle errors running the command
+  except proto.ScootException as e:
+    if re.search("No resources available", e) is None:
+      print(e)
+    elif re.search("Not started", e) > 0 or re.search("UNAVAILABLE", e) > 0:
+      raise docopt.DocoptExit("Scoot Daemon is not running!\n")
+    else:
+      raise docopt.DocoptExit("run request error: '{0}'".format(str(e)))
+  except Exception as e1:
+    raise docopt.DocoptExit("run request error:'{0}'".format(str(e1)))
+
+
+def poll(cmd):
+  #verify args
+  wait = cmd["--wait"]
+  if not wait:
+    wait = 0
+  try:
+    wait = int(wait)
+  except Exception as e:
+    raise docopt.DocoptExit("Wait must be an integer.")
+  runIds = cmd["<runId>"]
+  if not cmd["<runId>"]:
+    raise docopt.DocoptExit("RunId list is empty!")
```

Comment:
We have an awful lot of code here devoted to manual parsing and checking input options, types, etc. If there is not a way to handle the subcommand and required parameters automatically w/ docopt I am going to suggest using the standard argparse library.